sanitize models of phi

glenrs opened the pull request:
@mmastand as promised here is the first iteration of this issue. 

The only changes from master are in the `save_models` method. Below I have shown predictions before and after the model was saved and loaded. I will continue to work on more iterations on Thursday. Thanks for your feedback.

``` r
library(healthcareai)
#> healthcareai version 2.2.0
#> Please visit https://docs.healthcare.ai for full documentation and vignettes. Join the community at https://healthcare-ai.slack.com
library(devtools)
m <- machine_learn(pima_diabetes, patient_id, outcome = diabetes, 
                   models = "glm", tune = FALSE)
#> Training new data prep recipe...
#> Variable(s) ignored in prep_data won't be used to tune models: patient_id
#> 
#> diabetes looks categorical, so training classification algorithms.
#> 
#> After data processing, models are being trained on 12 features with 768 observations.
#> Based on n_folds = 5 and hyperparameter settings, the following number of models will be trained: 50 glm's
#> Training at fixed values: glmnet
#> 
#> *** Models successfully trained. The model object contains the training data minus ignored ID columns. ***
#> *** If there was PHI in training data, normal PHI protocols apply to the model object. ***
(predictions <- predict(m))
#> "predicted_diabetes" predicted by glmnet last trained: 2018-09-18 16:55:17
#> Performance in training: AUROC = 0.84
#> # A tibble: 768 x 11
#>    diabetes predicted_diabe… patient_id pregnancies plasma_glucose
#>  * <fct>               <dbl>      <int>       <int>          <int>
#>  1 Y                  0.756           1           6            148
#>  2 N                  0.0597          2           1             85
#>  3 Y                  0.476           3           8            183
#>  4 N                  0.0426          4           1             89
#>  5 Y                  0.675           5           0            137
#>  6 N                  0.180           6           5            116
#>  7 Y                  0.105           7           3             78
#>  8 N                  0.448           8          10            115
#>  9 Y                  0.887           9           2            197
#> 10 Y                  0.293          10           8            125
#> # ... with 758 more rows, and 6 more variables: diastolic_bp <int>,
#> #   skinfold <int>, insulin <int>, weight_class <chr>, pedigree <dbl>,
#> #   age <int>
predictions %>%
  plot()
```

![](https://i.imgur.com/NwplrLk.png)

``` r
save_models(m)
#> This model object does not contain PHI. If you want to save the data along with the model use sanitize_phi = FALSE
reloaded_m <- load_models("models.RDS")
(predictions <- predict(reloaded_m))
#> "predicted_diabetes" predicted by glmnet last trained: 2018-09-18 16:55:17
#> Performance in training: AUROC = 0.84
#> # A tibble: 768 x 11
#>    diabetes predicted_diabe… patient_id pregnancies plasma_glucose
#>  * <fct>               <dbl> <lgl>      <lgl>       <lgl>         
#>  1 Y                  0.756  NA         NA          NA            
#>  2 N                  0.0597 NA         NA          NA            
#>  3 Y                  0.476  NA         NA          NA            
#>  4 N                  0.0426 NA         NA          NA            
#>  5 Y                  0.675  NA         NA          NA            
#>  6 N                  0.180  NA         NA          NA            
#>  7 Y                  0.105  NA         NA          NA            
#>  8 N                  0.448  NA         NA          NA            
#>  9 Y                  0.887  NA         NA          NA            
#> 10 Y                  0.293  NA         NA          NA            
#> # ... with 758 more rows, and 6 more variables: diastolic_bp <lgl>,
#> #   skinfold <lgl>, insulin <lgl>, weight_class <lgl>, pedigree <lgl>,
#> #   age <lgl>
predictions %>%
  plot()
```

![](https://i.imgur.com/sBAUwLv.png)

Created on 2018-09-18 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

mmastand commented:
Thanks for getting this up, @glenrs! I think this is coming together. Upon further review and mental consolidation, I've had some design epiphanies. What do you think of these?
- First, I think we do want to get rid of that training data in the model object (not fill with NA). Part of the appeal to people removing the PHI might be simply to make the .RDS file smaller. So let's add a `select(target, predicted_target)` before saving. Additionally, the primary use case of a model that has been saved and loaded is predicting in production. And in that case. the data will be there after `predict`. There isn't a reason to need to explore or really do much else. What does this change break elsewhere?
- When save models is used with `sanitize=FALSE`, let's have it append a `PHI` to the file name.
- Each function that doesn't work without training data should give a consistent error. Rather than checking for training data in each function, would it be better to have `load_models` do it and flip an attribute, like `has_training_data`, which can be true or false?
- Should load models issue a warning about PHI if it is present?

Let's discuss these, make the changes we agree on, then take another look at how the downstream functions give warnings and errors.

glenrs commented:
Great ideas @mmastand! I will implement those asap. Just a couple questions first.

> So let's add a select(target, predicted_target) before saving.
- It is my understanding that these are not saved in the template. I believe the `target` and `predicted_target` are stored in a data frame object that is stored in `model_object[["model_name"]]$pred`. `target` and `predicted_target` are columns `obs` and `pred` respectively. For example, the `glmnet`'s `predicted_target` is stored at `model_object[["glmnet"]]$pred$pred` and the `target` is stored at `model_object[["glmnet"]]$pred$obs`. Now the question, so since you were suggesting to select only these two columns in the template data, would we like to set the template to NULL? If we set this object to NULL, only `explore` and `predict(m)` don't currently work. Predicting with new data still works `predict(m, new_data)`. I think it would also be an easy fix to get `predict(m)` to work since I think it just `cbinds` the template with the `predicted_target` and `target`. I could just have it return these two columns in `predict(m)`.

> When save models is used with sanitize=FALSE, let's have it append a PHI to the file name.
I really like this idea!

> Each function that doesn't work without training data should give a consistent error. Rather than checking for training data in each function, would it be better to have load_models do it and flip an attribute, like has_training_data, which can be true or false?
I agree that each function should list a consistent error that doesn't work. I was playing around with this on Tuesday. It would be very easy to just add the attribute to the model as we are saving it. However, if we are just setting the template to NULL, it is also very easy to just check whether the object is NULL. Is this okay? I am thinking that we would only need to check if it is NULL in 3 places `predict(m)`, `explore`, and `load_models`. Thoughts?

> Should load models issue a warning about PHI if it is present?
I could go either way. I will plan on issuing a warning though, unless you ask me to do otherwise. I think that users of the library might not completely understand how it works. We probably want to warn them of this issue, so they know what is in their model object. Hopefully they haven't already shared it like this, but it could stop them from sharing it in the future.

``` r
library(tidyverse)
library(healthcareai)
#> healthcareai version 2.2.0
#> Please visit https://docs.healthcare.ai for full documentation and vignettes. Join the community at https://healthcare-ai.slack.com

m <- machine_learn(pima_diabetes, patient_id, outcome = diabetes, models = c("glm", "rf"), tune = FALSE)
#> Training new data prep recipe...
#> Variable(s) ignored in prep_data won't be used to tune models: patient_id
#> 
#> diabetes looks categorical, so training classification algorithms.
#> 
#> After data processing, models are being trained on 12 features with 768 observations.
#> Based on n_folds = 5 and hyperparameter settings, the following number of models will be trained: 50 glm's and 5 rf's
#> Training at fixed values: glmnet
#> Training at fixed values: Random Forest
#> 
#> *** Models successfully trained. The model object contains the training data minus ignored ID columns. ***
#> *** If there was PHI in training data, normal PHI protocols apply to the model object. ***

attr(m, "recipe")$template <-
  NULL

interpret(m) %>%
  plot()
#> Warning in interpret(m): Interpreting glmnet model, but Random Forest
#> performed best in cross-validation and will be used to make predictions. To
#> use the glmnet model for predictions, extract it with x['glmnet'].
```

![](https://i.imgur.com/nR83fEW.png)

``` r
get_variable_importance(m) %>%
  plot()
```

![](https://i.imgur.com/XPjYhy0.png)

``` r
get_thresholds(m) %>%
  plot()
```

![](https://i.imgur.com/MQggPGn.png)

``` r
evaluate(m) 
#>      AUPR     AUROC 
#> 0.7033431 0.8393368

modifications <- list(weight_class = c("underweight", "normal", "overweight"),
                      plasma_glucose = c(75, 100),
                      diastolic_bp = 70)
pip(model = m, d = pima_diabetes[1:3, ], new_values = modifications)
#> # A tibble: 8 x 8
#>   patient_id variable original_value modified_value original_predic…
#>        <int> <chr>    <chr>          <chr>                     <dbl>
#> 1          1 plasma_… 148            75                       0.810 
#> 2          1 weight_… obese          underweight              0.810 
#> 3          1 diastol… 72             70                       0.810 
#> 4          2 weight_… overweight     normal                   0.0853
#> 5          2 plasma_… 85             75                       0.0853
#> 6          3 plasma_… 183            75                       0.722 
#> 7          3 weight_… normal         overweight               0.722 
#> 8          3 diastol… 64             70                       0.722 
#> # ... with 3 more variables: modified_prediction <dbl>, improvement <dbl>,
#> #   impact_rank <int>

predict(m, pima_diabetes)
#> Prepping data based on provided recipe
#> "predicted_diabetes" predicted by Random Forest last trained: 2018-09-20 07:48:06
#> Performance in training: AUROC = 0.84
#> # A tibble: 768 x 11
#>    diabetes predicted_diabe… patient_id pregnancies plasma_glucose
#>  * <fct>               <dbl>      <int>       <int>          <int>
#>  1 Y                  0.810           1           6            148
#>  2 N                  0.0853          2           1             85
#>  3 Y                  0.722           3           8            183
#>  4 N                  0.0256          4           1             89
#>  5 Y                  0.774           5           0            137
#>  6 N                  0.134           6           5            116
#>  7 Y                  0.585           7           3             78
#>  8 N                  0.265           8          10            115
#>  9 Y                  0.866           9           2            197
#> 10 Y                  0.759          10           8            125
#> # ... with 758 more rows, and 6 more variables: diastolic_bp <int>,
#> #   skinfold <int>, insulin <int>, weight_class <chr>, pedigree <dbl>,
#> #   age <int>

predict(m)
#> "predicted_diabetes" predicted by Random Forest last trained: 2018-09-20 07:48:06
#> Performance in training: AUROC = 0.84
#> # A tibble: 768 x 2
#>    diabetes predicted_diabetes
#>  * <fct>                 <dbl>
#>  1 Y                    0.667 
#>  2 N                    0.134 
#>  3 Y                    0.335 
#>  4 N                    0.0211
#>  5 Y                    0.551 
#>  6 N                    0.171 
#>  7 Y                    0.0989
#>  8 N                    0.485 
#>  9 Y                    0.732 
#> 10 Y                    0.380 
#> # ... with 758 more rows

explore(m)
#> Error in UseMethod("select_"): no applicable method for 'select_' applied to an object of class "NULL"
```

Created on 2018-09-20 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

glenrs commented:
@mmastand I made a mistake. `predict(m)` does actually work when the template is null. I am going to check all the other possible parameters to make sure it doesn't cause other problems.

glenrs commented:
I have implemented all those changes @mmastand that we have talked about! If something isn't clear, or we would like to make some changes, feel free to slack me! 

``` r
library(tidyverse)
library(healthcareai)
#> healthcareai version 2.2.0
#> Please visit https://docs.healthcare.ai for full documentation and vignettes. Join the community at https://healthcare-ai.slack.com

m <- machine_learn(pima_diabetes, patient_id, outcome = diabetes, models = c("glm", "rf"), tune = FALSE)
#> Training new data prep recipe...
#> Variable(s) ignored in prep_data won't be used to tune models: patient_id
#> 
#> diabetes looks categorical, so training classification algorithms.
#> 
#> After data processing, models are being trained on 12 features with 768 observations.
#> Based on n_folds = 5 and hyperparameter settings, the following number of models will be trained: 50 glm's and 5 rf's
#> Training at fixed values: glmnet
#> Training at fixed values: Random Forest
#> 
#> *** Models successfully trained. The model object contains the training data minus ignored ID columns. ***
#> *** If there was PHI in training data, normal PHI protocols apply to the model object. ***
save_models(m)
#> This model object does not contain PHI. If you want to save the data along with the model use sanitize_phi = FALSE
loaded_m <- load_models("models.RDS")

# Predict works as normal if the data is provided
predict(loaded_m, pima_diabetes)
#> Prepping data based on provided recipe
#> "predicted_diabetes" predicted by Random Forest last trained: 2018-09-20 14:13:26
#> Performance in training: AUROC = 0.84
#> # A tibble: 768 x 11
#>    diabetes predicted_diabe… patient_id pregnancies plasma_glucose
#>  * <fct>               <dbl>      <int>       <int>          <int>
#>  1 Y                  0.811           1           6            148
#>  2 N                  0.0653          2           1             85
#>  3 Y                  0.700           3           8            183
#>  4 N                  0.0235          4           1             89
#>  5 Y                  0.814           5           0            137
#>  6 N                  0.124           6           5            116
#>  7 Y                  0.562           7           3             78
#>  8 N                  0.238           8          10            115
#>  9 Y                  0.881           9           2            197
#> 10 Y                  0.763          10           8            125
#> # ... with 758 more rows, and 6 more variables: diastolic_bp <int>,
#> #   skinfold <int>, insulin <int>, weight_class <chr>, pedigree <dbl>,
#> #   age <int>
# Predict works if the data is not provided, but lists an informative message
predict(loaded_m)
#> "predicted_diabetes" predicted by Random Forest last trained: 2018-09-20 14:13:26
#> Performance in training: AUROC = 0.84
#> Your model was sanitized of PHI when stored and no data was provided. If this sanitation was in error, save your model with `sanitize_phi = False` in `save_models`. Otherwise, to include your data in this dataframe, please use `cbind` or `predict` with `newdata = [original_data]`.
#> # A tibble: 768 x 2
#>    diabetes predicted_diabetes
#>  * <fct>                 <dbl>
#>  1 Y                    0.652 
#>  2 N                    0.140 
#>  3 Y                    0.452 
#>  4 N                    0.0229
#>  5 Y                    0.493 
#>  6 N                    0.179 
#>  7 Y                    0.123 
#>  8 N                    0.436 
#>  9 Y                    0.665 
#> 10 Y                    0.472 
#> # ... with 758 more rows
# Explore lists a helpful error
explore(loaded_m)
#> Error in explore(loaded_m): When saving your model you specified that you wanted to remove your data, so there is not any data to explore with!

# Saving with data now creates the PHI tag, and all the functions work as normal
save_models(m, sanitize_phi = FALSE)
#> The model object being saved contains training data, minus ignored ID columns.
#> If there was PHI in training data, normal PHI protocols apply to the RDS file.
loaded_m_phi <- load_models("modelsPHI.RDS")
#> Your loaded model contains PHI! To remove PHI run `save_models(x)`.
predict(loaded_m_phi, pima_diabetes)
#> Prepping data based on provided recipe
#> "predicted_diabetes" predicted by Random Forest last trained: 2018-09-20 14:13:26
#> Performance in training: AUROC = 0.84
#> # A tibble: 768 x 11
#>    diabetes predicted_diabe… patient_id pregnancies plasma_glucose
#>  * <fct>               <dbl>      <int>       <int>          <int>
#>  1 Y                  0.811           1           6            148
#>  2 N                  0.0653          2           1             85
#>  3 Y                  0.700           3           8            183
#>  4 N                  0.0235          4           1             89
#>  5 Y                  0.814           5           0            137
#>  6 N                  0.124           6           5            116
#>  7 Y                  0.562           7           3             78
#>  8 N                  0.238           8          10            115
#>  9 Y                  0.881           9           2            197
#> 10 Y                  0.763          10           8            125
#> # ... with 758 more rows, and 6 more variables: diastolic_bp <int>,
#> #   skinfold <int>, insulin <int>, weight_class <chr>, pedigree <dbl>,
#> #   age <int>
predict(loaded_m_phi)
#> "predicted_diabetes" predicted by Random Forest last trained: 2018-09-20 14:13:26
#> Performance in training: AUROC = 0.84
#> # A tibble: 768 x 11
#>    diabetes predicted_diabe… patient_id pregnancies plasma_glucose
#>  * <fct>               <dbl>      <int>       <int>          <int>
#>  1 Y                  0.652           1           6            148
#>  2 N                  0.140           2           1             85
#>  3 Y                  0.452           3           8            183
#>  4 N                  0.0229          4           1             89
#>  5 Y                  0.493           5           0            137
#>  6 N                  0.179           6           5            116
#>  7 Y                  0.123           7           3             78
#>  8 N                  0.436           8          10            115
#>  9 Y                  0.665           9           2            197
#> 10 Y                  0.472          10           8            125
#> # ... with 758 more rows, and 6 more variables: diastolic_bp <int>,
#> #   skinfold <int>, insulin <int>, weight_class <chr>, pedigree <dbl>,
#> #   age <int>
explore(loaded_m_phi)
#> # A tibble: 625 x 9
#>    predicted_diabe… plasma_glucose   age weight_class pedigree pregnancies
#>  *            <dbl>          <dbl> <dbl> <chr>           <dbl>       <dbl>
#>  1            0.121             80    21 obese           0.140           3
#>  2            0.142             99    21 obese           0.140           3
#>  3            0.244            117    21 obese           0.140           3
#>  4            0.386            141    21 obese           0.140           3
#>  5            0.652            181    21 obese           0.140           3
#>  6            0.152             80    24 obese           0.140           3
#>  7            0.185             99    24 obese           0.140           3
#>  8            0.321            117    24 obese           0.140           3
#>  9            0.471            141    24 obese           0.140           3
#> 10            0.724            181    24 obese           0.140           3
#> # ... with 615 more rows, and 3 more variables: diastolic_bp <dbl>,
#> #   skinfold <dbl>, insulin <dbl>
```

Created on 2018-09-20 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

glenrs commented:
@mmastand don't worry about changing your mind! It isn't a big deal to remove that. I think that these new adjustments are great suggestions. Thank you for the improvements!

glenrs commented:
@mmastand I made those adjustments.

- I removed all warnings in save_models besides the original warning.
- I removed the PHI tag on file names.
- I added the new message for explore
- I changed the message in `load_models` to the one you provided. Thank you. **I have implemented this so that it only alerts the user if there is data in the template. This is not what you said. Is this okay? I can easy make this alert all the time. Let me know what you think.**

In addition:
- I fixed the message in print.predicted_df to have a more similar message to the new message for explore. **Thoughts?** This is the message that prints when the data doesn't exist in the model object.

``` r
library(tidyverse)
library(healthcareai)
#> healthcareai version 2.2.0
#> Please visit https://docs.healthcare.ai for full documentation and vignettes. Join the community at https://healthcare-ai.slack.com

m <- machine_learn(pima_diabetes, patient_id, outcome = diabetes, 
                   models = c("glm", "rf"), tune = FALSE)
#> Training new data prep recipe...
#> Variable(s) ignored in prep_data won't be used to tune models: patient_id
#> 
#> diabetes looks categorical, so training classification algorithms.
#> 
#> After data processing, models are being trained on 12 features with 768 observations.
#> Based on n_folds = 5 and hyperparameter settings, the following number of models will be trained: 50 glm's and 5 rf's
#> Training at fixed values: glmnet
#> Training at fixed values: Random Forest
#> 
#> *** Models successfully trained. The model object contains the training data minus ignored ID columns. ***
#> *** If there was PHI in training data, normal PHI protocols apply to the model object. ***
save_models(m)
loaded_m <- load_models("models.RDS")
predict(loaded_m, pima_diabetes)
#> Prepping data based on provided recipe
#> "predicted_diabetes" predicted by Random Forest last trained: 2018-09-25 09:13:23
#> Performance in training: AUROC = 0.84
#> # A tibble: 768 x 11
#>    diabetes predicted_diabe… patient_id pregnancies plasma_glucose
#>  * <fct>               <dbl>      <int>       <int>          <int>
#>  1 Y                  0.769           1           6            148
#>  2 N                  0.0634          2           1             85
#>  3 Y                  0.720           3           8            183
#>  4 N                  0.0156          4           1             89
#>  5 Y                  0.826           5           0            137
#>  6 N                  0.132           6           5            116
#>  7 Y                  0.584           7           3             78
#>  8 N                  0.272           8          10            115
#>  9 Y                  0.874           9           2            197
#> 10 Y                  0.721          10           8            125
#> # ... with 758 more rows, and 6 more variables: diastolic_bp <int>,
#> #   skinfold <int>, insulin <int>, weight_class <chr>, pedigree <dbl>,
#> #   age <int>
predict(loaded_m)
#> "predicted_diabetes" predicted by Random Forest last trained: 2018-09-25 09:13:23
#> Performance in training: AUROC = 0.84
#> Your model was sanitized of PHI when stored and no new data was provided. If this sanitation was in error, use `save_models(model, sanitize_phi = FALSE)` to keep it. Otherwise, to include your data in this dataframe, please use `cbind` or `predict(model, newdata)`.
#> # A tibble: 768 x 2
#>    diabetes predicted_diabetes
#>  * <fct>                 <dbl>
#>  1 Y                    0.751 
#>  2 N                    0.136 
#>  3 Y                    0.515 
#>  4 N                    0.0153
#>  5 Y                    0.493 
#>  6 N                    0.284 
#>  7 Y                    0.102 
#>  8 N                    0.443 
#>  9 Y                    0.725 
#> 10 Y                    0.453 
#> # ... with 758 more rows
explore(loaded_m)
#> Error in explore(loaded_m): Explore requires that data is attached to the model object. You must use `save_models(model, sanitize_phi = FALSE)` to keep it.

save_models(m, sanitize_phi = FALSE)
#> The model object being saved contains training data, minus ignored ID columns.
#> If there was PHI in training data, normal PHI protocols apply to the RDS file.
loaded_m_phi <- load_models("models.RDS")
#> *** If there was PHI in training data, normal PHI protocols apply to this model object. ***
predict(loaded_m_phi, pima_diabetes)
#> Prepping data based on provided recipe
#> "predicted_diabetes" predicted by Random Forest last trained: 2018-09-25 09:13:23
#> Performance in training: AUROC = 0.84
#> # A tibble: 768 x 11
#>    diabetes predicted_diabe… patient_id pregnancies plasma_glucose
#>  * <fct>               <dbl>      <int>       <int>          <int>
#>  1 Y                  0.769           1           6            148
#>  2 N                  0.0634          2           1             85
#>  3 Y                  0.720           3           8            183
#>  4 N                  0.0156          4           1             89
#>  5 Y                  0.826           5           0            137
#>  6 N                  0.132           6           5            116
#>  7 Y                  0.584           7           3             78
#>  8 N                  0.272           8          10            115
#>  9 Y                  0.874           9           2            197
#> 10 Y                  0.721          10           8            125
#> # ... with 758 more rows, and 6 more variables: diastolic_bp <int>,
#> #   skinfold <int>, insulin <int>, weight_class <chr>, pedigree <dbl>,
#> #   age <int>
predict(loaded_m_phi)
#> "predicted_diabetes" predicted by Random Forest last trained: 2018-09-25 09:13:23
#> Performance in training: AUROC = 0.84
#> # A tibble: 768 x 11
#>    diabetes predicted_diabe… patient_id pregnancies plasma_glucose
#>  * <fct>               <dbl>      <int>       <int>          <int>
#>  1 Y                  0.751           1           6            148
#>  2 N                  0.136           2           1             85
#>  3 Y                  0.515           3           8            183
#>  4 N                  0.0153          4           1             89
#>  5 Y                  0.493           5           0            137
#>  6 N                  0.284           6           5            116
#>  7 Y                  0.102           7           3             78
#>  8 N                  0.443           8          10            115
#>  9 Y                  0.725           9           2            197
#> 10 Y                  0.453          10           8            125
#> # ... with 758 more rows, and 6 more variables: diastolic_bp <int>,
#> #   skinfold <int>, insulin <int>, weight_class <chr>, pedigree <dbl>,
#> #   age <int>
explore(loaded_m_phi)
#> # A tibble: 625 x 9
#>    predicted_diabe… plasma_glucose   age weight_class pedigree pregnancies
#>  *            <dbl>          <dbl> <dbl> <chr>           <dbl>       <dbl>
#>  1            0.149             80    21 obese           0.140           3
#>  2            0.179             99    21 obese           0.140           3
#>  3            0.247            117    21 obese           0.140           3
#>  4            0.404            141    21 obese           0.140           3
#>  5            0.659            181    21 obese           0.140           3
#>  6            0.161             80    24 obese           0.140           3
#>  7            0.202             99    24 obese           0.140           3
#>  8            0.281            117    24 obese           0.140           3
#>  9            0.455            141    24 obese           0.140           3
#> 10            0.702            181    24 obese           0.140           3
#> # ... with 615 more rows, and 3 more variables: diastolic_bp <dbl>,
#> #   skinfold <dbl>, insulin <dbl>
```

Created on 2018-09-25 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).
